File: python/2024/06_guard_gallivant/aoc202406.py
"""AoC 6, 2024: Guard Gallivant."""

# Standard library imports
import pathlib
import sys

TURN_RIGHT = {(-1, 0): (0, 1), (0, 1): (1, 0), (1, 0): (0, -1), (0, -1): (-1, 0)}


def parse_data(puzzle_input):
    """Parse input."""
    return {
        (row, col): char
        for row, line in enumerate(puzzle_input.split("\n"))
        for col, char in enumerate(line)
    }


def part1(grid):
    """Solve part 1."""
    start = next(pos for pos, char in grid.items() if char == "^")
    return len({pos for pos, _ in walk(grid, start)})


def part2(grid):
    """Solve part 2.

    Put an obstacle at each point in the path from part 1. Start the walk right
    in front of the obstacle.
    """
    start = next(pos for pos, char in grid.items() if char == "^")

    # Find all the locations that can be reached, and how they are entered
    original_path = walk(grid, start)
    first_entries = {
        pos: (prev_pos, prev_dir)
        for (prev_pos, prev_dir), (pos, _) in zip(
            original_path[-2::-1], original_path[-1::-1]
        )
        if pos != start
    }

    # Add obstacles at all feasible positions
    obstacles = {
        obstacle
        for obstacle, (pos, dir) in first_entries.items()
        if not walk(grid | {obstacle: "#"}, pos, dir)
    }
    return len(obstacles)


def walk(grid, pos, dir=(-1, 0)):
    """Walk the grid until you walk off the map or get caught in a loop"""
    path = []
    seen = set()
    while pos in grid:
        if (pos, dir) in seen:  # Stuck in a loop
            return []
        path.append((pos, dir))
        seen.add((pos, dir))
        while grid.get(new_pos := (pos[0] + dir[0], pos[1] + dir[1])) == "#":
            dir = TURN_RIGHT[dir]
        pos = new_pos
    return path


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2024/06_guard_gallivant/aoc202406.ex
defmodule AOC2024.Day06 do
  @moduledoc """
  Advent of Code 2024, day 6: Guard Gallivant.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    grid =
      for {line, row} <- puzzle_input |> String.split("\n", trim: true) |> Enum.with_index(),
          {char, col} <- line |> String.split("", trim: true) |> Enum.with_index(),
          into: %{},
          do: {{row, col}, char}

    start = grid |> Map.filter(fn {_, char} -> char == "^" end) |> Map.keys() |> hd()
    {grid |> Map.put(start, "."), start}
  end

  @doc """
  Solve part 1.

  Currently off by one ...
  """
  def part1({grid, start}) do
    grid |> walk(start) |> MapSet.size()
  end

  @doc """
  Solve part 2.
  """
  def part2({grid, start}) do
    grid
    |> walk(start)
    |> MapSet.filter(&(&1 != start))
    |> Task.async_stream(fn obstacle -> walk(Map.put(grid, obstacle, "#"), start) end)
    |> Enum.reject(fn {:ok, result} -> result end)
    |> Enum.count()
  end

  @doc """
  Walk the grid from start until you exit the grid or find a loop.
  """
  def walk(grid, start), do: walk(grid, start, {-1, 0}, MapSet.new(), MapSet.new())

  def walk(grid, pos, dir, path, seen) do
    if MapSet.member?(seen, {pos, dir}), do: nil, else: walk_step(grid, pos, dir, path, seen)
  end

  defp walk_step(grid, pos = {row, col}, dir = {drow, dcol}, path, seen) do
    new_pos = {row + drow, col + dcol}

    if Map.has_key?(grid, new_pos) do
      case Map.get(grid, new_pos) do
        char when char in [".", "^"] ->
          walk(grid, new_pos, dir, MapSet.put(path, pos), MapSet.put(seen, {pos, dir}))

        "#" ->
          walk(grid, pos, {dcol, -drow}, path, seen)
      end
    else
      MapSet.put(path, pos)
    end
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
